mars/dataframe/expressions/arithmetic/core.py

```diff
@@ -0,0 +1,358 @@
+                                             index=left_chunk.index)
+                out_chunks.append(out_chunk)
+
+        return out_chunks
+
+    @classmethod
+    def _gen_out_chunks_with_all_shuffle(cls, op, out_shape, left, right):
+        out_chunks = []
+
+        # gen map chunks
+        reduce_chunks = [[], []]
+        for i in range(2):  # left, right
+            inp = left if i == 0 else right
+            map_chunks = []
+            for chunk in inp.chunks:
+                map_op = DataFrameIndexAlignMap(
+                    sparse=chunk.issparse(), index_shuffle_size=out_shape[0],
+                    column_shuffle_size=out_shape[1])
+                map_chunks.append(map_op.new_chunk([chunk], (np.nan, np.nan), index=chunk.index))
+
+            proxy_chunk = DataFrameShuffleProxy().new_chunk(map_chunks, ())
+            for out_idx in itertools.product(*(range(s) for s in out_shape)):
+                reduce_op = DataFrameIndexAlignReduce(i=out_idx,
+                                                      sparse=proxy_chunk.issparse(),
+                                                      shuffle_key=','.join(str(idx) for idx in out_idx))
+                reduce_chunks[i].append(
+                    reduce_op.new_chunk([proxy_chunk], (np.nan, np.nan), index=out_idx))
+
+        for left_chunk, right_chunk in zip(*reduce_chunks):
+            bin_op = op.copy().reset_key()
+            out_chunk = bin_op.new_chunk([left_chunk, right_chunk], (np.nan, np.nan),
+                                         index=left_chunk.index)
+            out_chunks.append(out_chunk)
+
+        return out_chunks
+
+    @classmethod
+    def _tile_both_dataframes(cls, op):
+        # if both of the inputs are DataFrames, axis is just ignored
+        left, right = op.inputs
+        nsplits = [[], []]
+        splits = [None, None]
+
+        # first, we decide the chunk size on each axis
+        # we perform the same logic for both index and columns
+        for axis, index_type in enumerate(['index_value', 'columns']):
+            # if both of the indexes are monotonic increasing or decreasing
+            left_chunk_index_min_max = cls._get_chunk_index_min_max(left, index_type, axis)
+            right_chunk_index_min_max = cls._get_chunk_index_min_max(right, index_type, axis)
+            if left_chunk_index_min_max is not None and right_chunk_index_min_max is not None:
+                # no need to do shuffle on this axis
+                left_splits, right_splits = split_monotonic_index_min_max(
+                    *(left_chunk_index_min_max + right_chunk_index_min_max))
+                left_increase = left_chunk_index_min_max[1]
+                right_increase = right_chunk_index_min_max[1]
+                splits[axis] = (left_splits, build_split_idx_to_origin_idx(left_splits, left_increase)), \
+                               (right_splits, build_split_idx_to_origin_idx(right_splits, right_increase))
```

Comment:
The tuple on the right along with the dict introduces three ``__getitem__``s for actual tiling functions, which introduces great difficulty for code readers when considering left and right ``splits`` are three-leveled tuples. Therefore I suggest using a namedtuple to pass these values to make this argument more clear.

Reply:
After refine, `splits` is a class instead of endless nested list.